File: Chapter06/Java/CommonControls/app/src/main/java/com/androidbook/commoncontrols/ListViewActivity4.java
package com.androidbook.commoncontrols;

<<<<<<< HEAD
import android.app.ListActivity;
=======
>>>>>>> afbbc0f (添加第6章源代码)
import android.content.ContentUris;
import android.database.Cursor;
import android.net.Uri;
import android.os.Bundle;
import android.provider.ContactsContract;
import android.util.Log;
import android.view.View;
import android.widget.ListView;
import android.widget.SimpleCursorAdapter;

<<<<<<< HEAD
public class ListViewActivity4 extends ListActivity {
=======
import androidx.appcompat.app.AppCompatActivity;

public class ListViewActivity4 extends AppCompatActivity {
>>>>>>> afbbc0f (添加第6章源代码)
    private static final String TAG = "ListViewActivity4";
    private static final Uri CONTACTS_URI = ContactsContract.Contacts.CONTENT_URI;
    private SimpleCursorAdapter adapter = null;
    private ListView lv = null;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.list);

<<<<<<< HEAD
        lv = getListView();
=======
        lv = findViewById(android.R.id.list);
>>>>>>> afbbc0f (添加第6章源代码)

        String[] projection = new String[]{ContactsContract.Contacts._ID,
                ContactsContract.Contacts.DISPLAY_NAME};

        Cursor c = getContentResolver().query(CONTACTS_URI,
                null, null, null, ContactsContract.Contacts.DISPLAY_NAME);

        String[] cols = new String[]{ContactsContract.Contacts.DISPLAY_NAME};
        int[] views = new int[]{android.R.id.text1};

        adapter = new SimpleCursorAdapter(this,
                android.R.layout.simple_list_item_multiple_choice,
                c, cols, views);

<<<<<<< HEAD
        this.setListAdapter(adapter);
=======
        lv.setAdapter(adapter);
>>>>>>> afbbc0f (添加第6章源代码)

        lv.setChoiceMode(ListView.CHOICE_MODE_MULTIPLE);
    }

    public void doClick(View view) {
        if (!adapter.hasStableIds()) {
            Log.v(TAG, "Data is not stable");
            return;
        }
        long[] viewItems = lv.getCheckedItemIds();
        for (int i = 0; i < viewItems.length; i++) {
            Uri selectedPerson = ContentUris.withAppendedId(
                    CONTACTS_URI, viewItems[i]);

            Log.v(TAG, selectedPerson.toString() + " is checked.");
        }
    }
}

File: Chapter06/Kotlin/CommonControls/app/src/main/java/com/androidbook/commoncontrols/ListViewActivity4.kt
package com.androidbook.commoncontrols

import android.content.ContentUris
import android.os.Bundle
import android.provider.ContactsContract
import android.util.Log
import android.view.View
import android.widget.ListView
import android.widget.SimpleCursorAdapter
import androidx.appcompat.app.AppCompatActivity

class ListViewActivity4: AppCompatActivity() {

    private lateinit var adapter: SimpleCursorAdapter
    private lateinit var lv: ListView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.list)

        lv = findViewById(android.R.id.list)

        val projection = arrayOf(
            ContactsContract.Contacts._ID,
            ContactsContract.Contacts.DISPLAY_NAME
        )

        val c = contentResolver.query(ContactsContract.Contacts.CONTENT_URI,
            null, null, null, ContactsContract.Contacts.DISPLAY_NAME)

        val cols = arrayOf(ContactsContract.Contacts.DISPLAY_NAME)
        val views = intArrayOf(android.R.id.text1)

        adapter = SimpleCursorAdapter(this,
            android.R.layout.simple_list_item_multiple_choice,
            c, cols, views
        )

        lv.adapter = adapter

        lv.choiceMode = ListView.CHOICE_MODE_MULTIPLE
    }

    public fun doClick(view: View?) {
        if(!adapter.hasStableIds()) {
            Log.v(TAG, "Data is not stable")
            return
        }
        val viewItems = lv.checkedItemIds
        for (i in viewItems.indices) {
            val selectedPerson = ContentUris.withAppendedId(
                ContactsContract.Contacts.CONTENT_URI, viewItems[i]
            )
            Log.v(TAG, "$selectedPerson is checked.")
        }
    }

    companion object {
        const val TAG = "ListViewActivity4"
    }
}